unittests/DXILLib/DxilLibTest.cpp

```diff
@@ -0,0 +1,152 @@
+struct HeapMalloc : public IMalloc {
+public:
+  ULONG STDMETHODCALLTYPE AddRef() override { return 1; }
+  ULONG STDMETHODCALLTYPE Release() override { return 1; }
+  STDMETHODIMP QueryInterface(REFIID iid, void **ppvObject) override {
+    return DoBasicQueryInterface<IMalloc>(this, iid, ppvObject);
+  }
+  void *STDMETHODCALLTYPE Alloc(SIZE_T cb) override {
+    return HeapAlloc(GetProcessHeap(), 0, cb);
+  }
+
+  void *STDMETHODCALLTYPE Realloc(void *pv, SIZE_T cb) override {
+    return HeapReAlloc(GetProcessHeap(), 0, pv, cb);
+  }
+
+  void STDMETHODCALLTYPE Free(void *pv) override {
+    HeapFree(GetProcessHeap(), 0, pv);
+  }
+
+  SIZE_T STDMETHODCALLTYPE GetSize(void *pv) override {
+    return HeapSize(GetProcessHeap(), 0, pv);
+  }
+
+  int STDMETHODCALLTYPE DidAlloc(void *pv) override {
+    return -1; // don't know
+  }
+
+  void STDMETHODCALLTYPE HeapMinimize(void) override {}
+};
+
+// Just need to compile something simple to verify things about its output
+void CompileSomething(dxc::DxcDllSupport &compilerLib,
+                      IDxcOperationResult **outResult) {
+
+  CComPtr<IDxcCompiler> compiler;
+  CComPtr<IDxcLibrary> library;
+  CComPtr<IDxcOperationResult> result;
+  const char *source = "float4 main() : SV_Target { return 1; }";
+  ASSERT_TRUE(
+      SUCCEEDED(compilerLib.CreateInstance(CLSID_DxcCompiler, &compiler)));
+  ASSERT_TRUE(
+      SUCCEEDED(compilerLib.CreateInstance(CLSID_DxcLibrary, &library)));
+
+  CComPtr<IDxcBlobEncoding> sourceBlob;
+  IFT(library->CreateBlobWithEncodingOnHeapCopy(source, strlen(source), CP_UTF8,
+                                                &sourceBlob));
+
+  ASSERT_TRUE(
+      SUCCEEDED(compiler->Compile(sourceBlob, L"hlsl.hlsl", L"main", L"ps_6_0",
+                                  nullptr, 0, nullptr, 0, nullptr, &result)));
+  HRESULT status;
+  ASSERT_TRUE(SUCCEEDED(result->GetStatus(&status)));
+  ASSERT_TRUE(SUCCEEDED(status));
+  *outResult = result.Detach();
+}
+
+// Check that dxil lib can be loaded by default when creating
+// the compiler library. If the external DXIL Library isn't found, a warning is
+// issued. Test that there is no such warning.
+TEST(DxilLibTest, LoadFromCompiler) {
+
+  // Preloading validator lib so dxcompiler.dll will find it without RPATH's
+  // help Only Linux needs this, but we <3 Linux, so we help it out sometimes
+  dxc::DxcDllSupport validatorLib;
+  ASSERT_TRUE(SUCCEEDED(
+      validatorLib.InitializeForDll(dxc::kDxilLib, "DxcCreateInstance")))
+      << "Couldn't find " << dxc::kDxilLib;
+
+  dxc::DxcDllSupport compilerLib;
+  ASSERT_TRUE(SUCCEEDED(compilerLib.Initialize()));
+
+  // Compile trivial, valid shader and check for warnings
+  CComPtr<IDxcOperationResult> result;
+  CompileSomething(compilerLib, &result);
+
+  CComPtr<IDxcBlobEncoding> errBuf;
+  ASSERT_TRUE(SUCCEEDED(result->GetErrorBuffer(&errBuf)));
+
+  EXPECT_EQ(errBuf->GetBufferSize(), 0U)
+      << "Unexpected warning found in compiler error buffer: "
```

Comment:
Could this hit error instead of warning?

Reply:
I guess it's possible, but given how simple the only shader that is compiled here is, it would indicate a serious problem in the compiler.

If this is to fail, the diagnostic I expect to find is the one that says you don't have a dxil lib and so your shaders won't be signed. This dxil lib won't sign your shader libs, but it avoids this warning. If we should get some other warning or error in this buffer, that would be unexpected, but this would catch it.